exchanges/gateio/gateio.go

```diff
@@ -448,145 +1041,2947 @@ func (g *Gateio) SendAuthenticatedHTTPRequest(ctx context.Context, ep exchange.U
+	if arg.PreviewID == "" {
+		return nil, errMissingPreviewID
+	}
+	if arg.BuyCurrency.IsEmpty() {
+		return nil, fmt.Errorf("%v, buy currency can not empty", errInvalidCurrency)
+	}
+	if arg.SellCurrency.IsEmpty() {
+		return nil, fmt.Errorf("%v, sell currency can not empty", errInvalidCurrency)
+	}
+	if !((arg.SellAmount >= 1 && arg.SellAmount <= 10000) || arg.SellAmount == 0) {
+		return nil, fmt.Errorf("%v, sell_amount must greater than 1 less than 10000", errInvalidAmount)
+	}
+	if !((arg.BuyAmount >= 0.0001 && arg.BuyAmount <= 0.1) || arg.BuyAmount == 0) {
+		return nil, fmt.Errorf("%v, buy_amount must greater than 0.0001 less than 0.1", errInvalidAmount)
+	}
+	var response InitFlashSwapOrderPreviewResponse
+	return &response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodPost, flashSwapOrdersPreview, nil, &arg, &response)
+}
+
+// IsValidPairString returns true if the string represents a valid currency pair
+func (g *Gateio) IsValidPairString(currencyPair string) bool {
+	if len(currencyPair) < 3 {
+		return false
+	}
+	if strings.Contains(currencyPair, g.CurrencyPairs.RequestFormat.Delimiter) {
+		result := strings.Split(currencyPair, g.CurrencyPairs.RequestFormat.Delimiter)
+		return len(result) >= 2
+	}
+	return false
+}
+
+// ********************************* Trading Fee calculation ********************************
+
+// GetFee returns an estimate of fee based on type of transaction
+func (g *Gateio) GetFee(ctx context.Context, feeBuilder *exchange.FeeBuilder) (fee float64, err error) {
+	switch feeBuilder.FeeType {
+	case exchange.CryptocurrencyTradeFee:
+		feePairs, err := g.GetPersonalTradingFee(ctx, feeBuilder.Pair)
+		if err != nil {
+			return 0, err
+		}
+		if feeBuilder.IsMaker {
+			fee = calculateTradingFee(feePairs.MakerFee,
+				feeBuilder.PurchasePrice,
+				feeBuilder.Amount)
+		} else {
+			fee = calculateTradingFee(feePairs.TakerFee,
+				feeBuilder.PurchasePrice,
+				feeBuilder.Amount)
+		}
+	case exchange.CryptocurrencyWithdrawalFee:
+		fee = getCryptocurrencyWithdrawalFee(feeBuilder.Pair.Base)
+	case exchange.OfflineTradeFee:
+		fee = getOfflineTradeFee(feeBuilder.PurchasePrice, feeBuilder.Amount)
+	}
+
+	if fee < 0 {
+		fee = 0
+	}
+
+	return fee, nil
+}
+
+// getOfflineTradeFee calculates the worst case-scenario trading fee
+func getOfflineTradeFee(price, amount float64) float64 {
+	return 0.002 * price * amount
+}
+
+func calculateTradingFee(feeForPair, purchasePrice, amount float64) float64 {
+	return (feeForPair / 100) * purchasePrice * amount
+}
+
+func getCryptocurrencyWithdrawalFee(c currency.Code) float64 {
+	return WithdrawalFees[c]
+}
+
+// determineAccount determine which account is enabled by sending get order requests and return empty asset item with error message
+// to help determine the account type enabled by the credential.
+func (g *Gateio) determineAccount(ctx context.Context) (asset.Item, error) {
```

Comment:
😆 Oh dear

Reply:
It's embarrassing😅. But, I was thinking of connecting to the enabled account web socket endpoint automatically. and, I am gonna delete it now.